File: python/algorithms/quick_sort.py
#!/usr/bin/env python


def quick_sort(iterable):
  """Sorts the iterable.

  Quicksort sort is a divide and conquer, comparison sorting algorithm.
  Conceptually, the steps are:
  * Select a pivot from the collection.
  * Partition the collection so that all the items less than the pivot come
    before the pivot and all the items greater than the pivot come after the
    pivot.
  * Repeatedly apply these steps to the collection of smaller items and the
    collection of larger items until the size of the collections are either zero
    or one, which are sorted by definition.

  The pivot selection and partitioning logic can be implemented in several
  different ways and greatly affect performance. This implementation uses
  Hoare's original partition scheme and the middle index as a pivot.

  * O(n log n) time complexity
  * O(log n) space complexity
  * not stable
  * not adaptive

  Args:
    iterable: A collection that is iterable.
  Raises:
    TypeError: If iterable is not iterable.
  """
  try:
    _ = iter(iterable)
  except TypeError:
    raise TypeError('\'{}\' object is not iterable'.format(
        type(iterable).__name__))

  def _partition(iterable, low, high):
    middle = int(low + (high - low) / 2)
    pivot = iterable[middle]
    i = low - 1
    j = high + 1
    while True:
      i += 1
      while iterable[i] < pivot:
        i += 1
      j -= 1
      while iterable[j] > pivot:
        j -= 1
      if i >= j:
        return j
      iterable[i], iterable[j] = iterable[j], iterable[i]

  def _quick_sort(iterable, low, high):
    if low >= high:
      return
    index = _partition(iterable, low, high)
    _quick_sort(iterable, low, index)
    _quick_sort(iterable, index + 1, high)

  _quick_sort(iterable, 0, len(iterable) - 1)

File: c/algorithms/quick_sort.c
#include "quick_sort.h"

static void swap(int* i, int* j) {
  int temp = *i;
  *i = *j;
  *j = temp;
}

static size_t partition(int array[], size_t low, size_t high) {
  size_t middle = low + (high - low) / 2;
  int pivot = array[middle];
  size_t i = low - 1;
  size_t j = high + 1;
  while (1) {
    do {
      ++i;
    } while (array[i] < pivot);
    do {
      --j;
    } while (array[j] > pivot);
    if (i >= j) {
      return j;
    }
    swap(array + i, array + j);
  }
}

static void quick_sort_internal(int array[], size_t low, size_t high) {
  if (low >= high) {
    return;
  }
  size_t index = partition(array, low, high);
  quick_sort_internal(array, low, index);
  quick_sort_internal(array, index + 1, high);
}

void quick_sort(int array[], size_t length) {
  if (length == 0) {
    return;
  }

  quick_sort_internal(array, 0, length - 1);
}
